### PR TITLE
Google oauth2

### PR DESCRIPTION
**Pull request by Olha98**

### CODE REVIEW

File: backend/package.json [hunk 1]
```diff
@@ -30,18 +36,24 @@
...
     "mongoose": "^5.10.14",
     "mongoose-paginate-v2": "^1.3.11",
     "morgan": "^1.10.0",
+    "passport": "^0.4.1",
+    "passport-google-oauth2": "^0.2.0",
+    "passport-local": "^1.0.0",
     "node-cron": "^2.0.3",
     "should": "^13.2.3",
     "sinon": "^9.2.1",
+    "uuid4": "^2.0.2",
```
Comment (alucardpetrovich) [inline backend/package.json]: I guess, we don't need it anymore

File: backend/src/api/auth/auth.controller.js [hunk 2]
```diff
@@ -59,3 +62,53 @@ exports.loginUser = async (req, res, next) => {
     token: token,
   });
 };
+
+exports.authorize = async (req, res, next) => {
```
Comment (alucardpetrovich) [inline backend/src/api/auth/auth.controller.js]: But... why? We already have authorization middleware in utils folder. Please be careful next time and don't do unnesesary work

File: backend/src/api/auth/auth.routers.js [hunk 3]
```diff
@@ -28,4 +30,28 @@ authRouter.post(
...
+authRouter.get(
+  '/google',
+  passport.authenticate('google', { scope: ['email', 'profile'] }),
+);
+
+authRouter.get(
+  '/google/callback',
+
+  passport.authenticate('google', { session: false }),
+  (req, res, next) => {
```
Comment (alucardpetrovich) [inline backend/src/api/auth/auth.routers.js]: First of all, this middleware should be written in controller. Second, what's the reason behind placing another middleware on 52 row? It won't work under any circumstances

File: backend/src/api/auth/strategies/google.strategy.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+  passport.use(
+    new GoogleStrategy(
+      {
+        clientID: process.env.GOOGLE_OAUTH_CLIENT_ID,
+        clientSecret: process.env.GOOGLE_OAUTH_CLIENT_SECRET,
+        callbackURL: process.env.GOOGLE_OAUTH_CALLBACK_URL,
+        passReqToCallback: true,
+      },
+      async (request, accessToken, refreshToken, profile, done) => {
+        console.log(profile, ">>>profile<<<<")
```
Comment (alucardpetrovich) [inline backend/src/api/auth/strategies/google.strategy.js]: I guess, this logs could be removed now

File: backend/src/api/session/session.model.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
+const mongoose = require("mongoose");
```
Comment (alucardpetrovich) [inline backend/src/api/session/session.model.js]: We don't need session model. We've already discussed, that we're saving tokens in user model

File: backend/src/api/auth/auth.routers.js [hunk 6]
```diff
@@ -28,4 +30,28 @@ authRouter.post(
...
+    const sessionToken = jwt.sign(
```
Comment (alucardpetrovich) [inline backend/src/api/auth/auth.routers.js]: Third, you need to save token into user document. Just like with regular sign-in

File: backend/src/server.js [hunk 7]
```diff
@@ -1,13 +1,18 @@
...
-const path = require('path');
-const morgan = require('morgan');
-const authRouter = require('./api/auth/auth.routers');
-const usersRouter = require('./api/users/user.router');
-const transactionRouter = require('./api/transactions/transactionRouter');
-require('dotenv').config({ path: path.join('./.env') });
-
-const AppError = require('./api/errors/appError');
+const express = require("express");
+const cors = require("cors");
```
Comment (alucardpetrovich) [inline backend/src/server.js]: Prettier? Why we have different quotes here?

File: backend/src/server.js [hunk 8]
```diff
@@ -16,9 +21,12 @@ const { updateInfo } = require('./api/cron/cronUpdateInfo');
 
 const giftsRouter = require('./api/gifts/gifts.router');
 
+
+
 class CrudServer {
   constructor() {
     this.server = null;
+    // this.config = config;
```
Comment (alucardpetrovich) [inline backend/src/server.js]: Should be removed completely

File: backend/src/server.js [hunk 9]
```diff
@@ -30,13 +38,6 @@ class CrudServer {
     this.startListening();
   }
 
-  async startForTest() {
```
Comment (alucardpetrovich) [inline backend/src/server.js]: Why you've removed it??? All auto-tests would fail now because of that!!!

File: backend/src/server.js [hunk 10]
```diff
@@ -60,9 +61,28 @@ class CrudServer {
   }
 
   initMiddlewares() {
+    this.server.use(cookieParser());
+    this.server.use(passport.initialize());
+    this.server.use(passport.session());
```
Comment (alucardpetrovich) [inline backend/src/server.js]: You don't need passport.session, deserializeUser, serializeUser at all

Comment (Olha98) [reply backend/src/server.js]: ok

File: backend/src/server.js (hunk 9 above)
Comment (Olha98) [reply backend/src/server.js]: ok

File: backend/src/server.js (hunk 8 above)
Comment (Olha98) [reply backend/src/server.js]: ok

File: backend/src/server.js (hunk 7 above)
Comment (Olha98) [reply backend/src/server.js]: ok

File: backend/src/api/auth/strategies/google.strategy.js (hunk 4 above)
Comment (Olha98) [reply backend/src/api/auth/strategies/google.strategy.js]: ok

File: backend/package.json (hunk 1 above)
Comment (Olha98) [reply backend/package.json]: ok

File: backend/src/api/auth/auth.controller.js (hunk 2 above)
Comment (Olha98) [reply backend/src/api/auth/auth.controller.js]: ok

File: backend/src/api/auth/auth.routers.js (hunk 3 above)
Comment (Olha98) [reply backend/src/api/auth/auth.routers.js]: ok

File: backend/src/api/auth/auth.routers.js (hunk 6 above)
Comment (Olha98) [reply backend/src/api/auth/auth.routers.js]: ok

File: backend/src/api/auth/auth.controller.js [hunk 11]
```diff
@@ -74,3 +74,22 @@ exports.logoutUser = async (req, res, next) => {
 function removeExpiredTokens(array) {
   return array.filter(item => item.expires > Date.now());
 }
+
+exports.AuthGoogle = async (req, res, next) => {
+  const user = req.user;
+  const existingUser = await UserModel.findById(user._id);
```
Comment (alucardpetrovich) [inline backend/src/api/auth/auth.controller.js]: I think, we could just simply use req.user. We don't need to fetch user one more time from DB

File: backend/src/api/auth/auth.controller.js [hunk 12]
```diff
@@ -74,3 +74,22 @@ exports.logoutUser = async (req, res, next) => {
...
+  const expiresIn = 2 * 24 * 60 * 60;
```
Comment (alucardpetrovich) [inline backend/src/api/auth/auth.controller.js:78]: could you move it to env variable or sort of a constant at least? This value the same for regular sign-in and google sign-in

File: backend/src/api/auth/auth.controller.js [hunk 13]
```diff
@@ -74,3 +74,22 @@ exports.logoutUser = async (req, res, next) => {
...
+    expiresIn,
+  });
+
+  existingUser.tokens.push({ token, expires: Date.now() + expiresIn });
+  existingUser.tokens = removeExpiredTokens(existingUser.tokens);
+  await existingUser.save();
+  return res.redirect(`http://localhost:3000?token=${token}`);
+};
+
+exports.AuthGoogleCallbackErr = async (req, res, next) => {
```
Comment (alucardpetrovich) [inline backend/src/api/auth/auth.controller.js]: The same comment as I left previously. This middleware won't be executed under any circumstances, meaning your user would never be redirected to `http://localhost:3000/notfound `

File: backend/src/server.js (hunk 10 above)
Comment (alucardpetrovich) [reply backend/src/server.js]: this issue still not resolved. You still have `.serializeUser` and `.deserializeUser` in your code

Comment (Olha98) [reply backend/src/server.js]: ok

File: backend/src/api/auth/auth.controller.js (hunk 11 above)
Comment (Olha98) [reply backend/src/api/auth/auth.controller.js]: ok

File: backend/src/api/auth/auth.controller.js (hunk 12 above)
Comment (Olha98) [reply backend/src/api/auth/auth.controller.js:78]: ok

File: backend/src/api/auth/auth.controller.js (hunk 13 above)
Comment (Olha98) [reply backend/src/api/auth/auth.controller.js]: ок

**Review by alucardpetrovich — APPROVED** [review]

**Pull request merged**